cajita/unit_test/tstParticleGridDistributor3d.hpp

```diff
@@ -0,0 +1,406 @@
+
+                // Increment the particle count.
+                ++pid;
+            }
+
+    // Copy to the device space.
+    auto particles_mirror =
+        Cabana::create_mirror_view_and_copy( TEST_DEVICE(), particles );
+
+    // Redistribute the particles.
+    auto coords_mirror = Cabana::slice<0>( particles_mirror, "coords" );
+    Cajita::particleGridMigrate( *block, coords_mirror, particles_mirror, 0,
+                                 true );
+
+    // Copy back to check.
+    particles = Cabana::create_mirror_view_and_copy( Kokkos::HostSpace(),
+                                                     particles_mirror );
+    coords = Cabana::slice<0>( particles, "coords" );
+    linear_ids = Cabana::slice<1>( particles, "linear_ids" );
+
+    // Check that we got as many particles as we should have.
+    EXPECT_EQ( coords.size(), num_particle );
+    EXPECT_EQ( linear_ids.size(), num_particle );
+
+    // Check that all of the particle ids are equal to this rank id.
+    for ( int p = 0; p < num_particle; ++p )
+        EXPECT_EQ( linear_ids( p ), global_grid->blockId() );
+
+    // Check that all of the particles are now in the local domain.
+    double local_low[3] = { local_mesh.lowCorner( Cajita::Own(), Dim::I ),
+                            local_mesh.lowCorner( Cajita::Own(), Dim::J ),
+                            local_mesh.lowCorner( Cajita::Own(), Dim::K ) };
+    double local_high[3] = { local_mesh.highCorner( Cajita::Own(), Dim::I ),
+                             local_mesh.highCorner( Cajita::Own(), Dim::J ),
+                             local_mesh.highCorner( Cajita::Own(), Dim::K ) };
+    for ( int p = 0; p < num_particle; ++p )
+        for ( int d = 0; d < 3; ++d )
+        {
+            EXPECT_GE( coords( p, d ), local_low[d] );
+            EXPECT_LE( coords( p, d ), local_high[d] );
+        }
+}
+
+auto createGrid( const Cajita::ManualPartitioner& partitioner,
+                 const std::array<bool, 3>& is_periodic,
+                 const double cell_size )
+{
+    // Create the global grid.
+    std::array<int, 3> global_num_cell = { 18, 15, 9 };
+    std::array<double, 3> global_low_corner = { 1.2, 3.3, -2.8 };
+    std::array<double, 3> global_high_corner = {
+        global_low_corner[0] + cell_size * global_num_cell[0],
+        global_low_corner[1] + cell_size * global_num_cell[1],
+        global_low_corner[2] + cell_size * global_num_cell[2] };
+    auto global_mesh = Cajita::createUniformGlobalMesh(
+        global_low_corner, global_high_corner, global_num_cell );
+    auto global_grid = Cajita::createGlobalGrid( MPI_COMM_WORLD, global_mesh,
+                                                 is_periodic, partitioner );
+    return global_grid;
+}
+
+//---------------------------------------------------------------------------//
+// RUN TESTS
+//---------------------------------------------------------------------------//
+TEST( TEST_CATEGORY, not_periodic_test )
```

Comment:
Same here with overlap - these two could probably be combined mostly as a single function (still want separate functions at GTest level though for delineation)

Reply:
Fixed